gap/pairingheap.gi

```diff
@@ -101,19 +92,60 @@ function(heap)
         fi;
     end;
 
-    merge_pairs := function(isLess, heaps)
-        local h, res;
+    # TODO: Find out whether recursive vs non-recursive
+    #       has any measurable benefits
+
+    # Non-recursive merge-pairs
+    merge_pairs_nr := function(isLess, heaps)
+        local l, res, tmp, k, s, i, r;
 
-        if Length(heaps) = 0 then
+        l := Length(heaps);
+
+        if l = 0 then
             return [0,0,0];
+        elif l = 1 then
+            return heaps[1];
         else
-            res := heaps[1];
-
-            for h in heaps{[2..Length(heaps)]} do
-                res := meld(isLess, res, h);
+            res := heaps;
+            k := l;
+            s := 1;
+
+            while k > 1 do
+                r := RemInt(k, 2);
+                k := QuoInt(k, 2);
+                s := 2*s;
+
+                for i in [1..k] do
+                    res[s*i] := meld(isLess, res[s*i - s/2], res[s*i]);
```

Comment:
This code is really inefficient. I rewrote it as follows, and now the non-recursive version is faster than the recursive one:
```gap
            while k > 1 do
                r := RemInt(k, 2);
                k := QuoInt(k, 2);
                old_s := s;
                s := 2*s;

                for i in [s, 2*s .. k*s] do
                    res[i] := meld(isLess, res[i - old_s], res[i]);
                od;
                if r = 1 then
                    i := i + s;
                    res[i] := res[i - old_s];
                    k := k + 1;
                fi;
            od;
            return res[i];
```